chainermn/communicators/pure_nccl_communicator.py

```diff
@@ -106,58 +111,41 @@ def _allreduce_grad_async(self, model, stream):
+
+        # unpack params from buffer A -> params
+        self._unpack_params_from_buffer(params,
                                         allreduce_grad_dtype, n_elems, stream)
 
-    def _assign_for_allreduce_grad(self, grad_dtype, allreduce_grad_dtype,
-                                   n_elems):
+    def _prepare_allreduce_pack_buffer(self, grad_dtype, allreduce_grad_dtype,
+                                       n_elems):
         allreduce_grad_n_bytes = allreduce_grad_dtype.itemsize * n_elems
         needs_sync = False
+
         if self.gpu_buffer_a.size != allreduce_grad_n_bytes:
             self.gpu_buffer_a.assign(allreduce_grad_n_bytes)
             needs_sync = True
         if self.gpu_buffer_b.size != allreduce_grad_n_bytes:
             self.gpu_buffer_b.assign(allreduce_grad_n_bytes)
             needs_sync = True
 
-        if grad_dtype != allreduce_grad_dtype:
-            grad_n_bytes = grad_dtype.itemsize * n_elems
-            if self.gpu_tmp_buffer.size != grad_n_bytes:
-                self.gpu_tmp_buffer.assign(grad_n_bytes)
-                needs_sync = True
         return needs_sync
 
-    def _pack_params_to_buffer(self, params, grad_dtype, allreduce_grad_dtype,
+    def _pack_params_to_buffer(self, params, allreduce_grad_dtype,
                                n_elems, stream):
         if self.batched_copy:
             params_data = _ParamsData(params, 'grad')
             _batched_pack_params(params_data, self.gpu_buffer_a,
                                  allreduce_grad_dtype)
             self.params_data = params_data
             # self.params_data will be re-used by _unpack_params_from_buffer
-            return
-        if grad_dtype == allreduce_grad_dtype:
-            _memory_utility.pack_params(
-                params, grad_dtype.itemsize, 'grad',
-                self.gpu_buffer_a, stream=stream, transfer_dtype=grad_dtype)
         else:
-            if self.grad_dtype_to_allreduce_dtype_kernel is None:
-                self.grad_dtype_to_allreduce_dtype_kernel = \
-                    _get_converting_kernel(
-                        grad_dtype, allreduce_grad_dtype,
-                        'grad_dtype_to_allreduce_dtype_kernel')
-
             _memory_utility.pack_params(
-                params, grad_dtype.itemsize, 'grad',
-                self.gpu_tmp_buffer, stream=stream, transfer_dtype=grad_dtype)
-
-            self.grad_dtype_to_allreduce_dtype_kernel(
-                self.gpu_tmp_buffer.array(n_elems, dtype=grad_dtype),
-                self.gpu_buffer_a.array(n_elems,
-                                        dtype=allreduce_grad_dtype),
+                params, 'grad',
+                self.gpu_buffer_a,
+                transfer_dtype=allreduce_grad_dtype,
                 stream=stream)
 
-    def _unpack_params_from_buffer(self, params, grad_dtype,
+    def _unpack_params_from_buffer(self, params,
                                    allreduce_grad_dtype, n_elems, stream):
```

Comment:
How about delete `n_elems` because it is not used.

Reply:
removed! thanks